dbt/include/global_project/macros/materializations/helpers.sql

```diff
@@ -1,5 +1,10 @@
 {% macro run_hooks(hooks, inside_transaction=True) %}
   {% for hook in hooks | selectattr('transaction', 'equalto', inside_transaction)  %}
+    {% if not inside_transaction and loop.first %}
+      {% call statement(auto_begin=inside_transaction) %}
+        commit;
+      {% endcall %}
```

Comment:
the more time passes, the more i don't like our implementation of begin/commit in adapters. it's jarring to see an explicit commit query issued here, i'd expect to see `adapter.commit()`. but, we can't do that because commit raises an exception if no transaction is opened, etc. etc. etc.

this is a reasonable solution given the constraints that we have now, but rebooting our implementation of these concepts in the adapter could fix a lot of downstream issues. naively:

- begin() and commit() should mimic the behavior of issuing those queries on the underlying database (i.e. should not raise exceptions if a transaction is / is not open)
- we should implement additional adapter functions to interrogate the connection to find out if we are in a transaction or not. for example, in postgres, we can issue `select txid_current();` twice and if the value is the same, we're in a transaction.

Reply:
Totally agree @cmcarthur -- I do like being explicit about where the begin/commit occurs, but I think our current approach of sprinkling these statements in to fix bugs is not a good or sustainable one. Let's have a think about this over here https://github.com/fishtown-analytics/dbt/issues/629